### ISSUE TITLE
Create NVDA's com object as NVDA Controller library equivalent

### ISSUE THREAD

**Issue by tseykovets** [nvaccess/nvda#13526]
Consider creating an NVDA's com object (OLE Automation) that will register with the system when NVDA is running. This com object should provide an API for interacting with NVDA. In particular, the speak text and the output text on the braille display.

There is currently the [NVDA Controller library](http://www.nvda-project.org/nvdaControllerClient/nvdaControllerClient_20100219.7z). It allows to implement this functionality. However, the implementation based on the com object seems to be more convenient.

The library should be included in the application or framework that needs to use the API. Often this is difficult or impossible due to the licensing policy of the application or framework. Including an external component with a separate license requires complex negotiations and is not always possible. To date, developers of the accessibility layer in the Open JDK have already encountered such a problem.

If NVDA, at startup, registers a com-object with an API in the system for interacting with NVDA, then any application or framework will be able to implement the necessary methods for interacting with NVDA. For example, methods for sending announcement messages.

**Comment by feerrenrut** [comment]
>Often this is difficult or impossible due to the licensing policy of the application or framework. Including an external component with a separate license requires complex negotiations and is not always possible. To date, developers of the accessibility layer in the Open JDK have already encountered such a problem.

I'm not sure the solution described here will address this particular problem. It would be helpful if you could elaborate on this licensing problem further.

There are other benefits to a COM API, but for us this isn't a high priority because generally the NVDA controller library approach seems to cover the usage requirements we are aware of.

**Comment by feerrenrut** [comment]
@savoptik You'll need to explain how this provides value for NVDA users generally. If you aren't sure how to do this, start with describing one concrete use-case that helps NVDA users.

Otherwise, I'd suggest building your own communication layer to achieve this.

If your company would like to consult with NV Access about this work, please contact info@nvaccess.org

**Comment by leonardder** [comment]
Creating a COM object isn't that trivial and breaks backwards compatibility in a major way. We'd still need to support the old API. Also, having to register the COM object imposes some challenges with portable copies. I'd be very reluctant to do this.

**Comment by OzancanKaratas** [comment]
@leonardder, potentially, we register the COM object when NVDA is started and unregister it when it is closed. Can this option work on a portable copy?

**Comment by leonardder** [comment]
It is technically possible, but even on installed copies, every client will have to recreate the com object when NVDA is restarted. I consider that a major drawback of the approach.

**Comment by Adriani90** [comment]
@OzancanKaratas what would this mean in practice? Would the user have to copy in NVDA every controlerClient DLL from every application that calls it? This would be an overload in my view.

@seanbudd, @gerald-hartig what do you think?

**Comment by OzancanKaratas** [comment]
No, this means that any developer can use the DLL file in the NVDA directory. Examples include Visual C++ packages. Even if a Visual C++ package isn't distributed with the program, the Visual C++ installed on the system can be used.
